### ISSUE TITLE
Change the default `type: any` to `type: $any`

### ISSUE THREAD

**Issue by popematt** [amzn/ion-schema#58]
**Problem**

From the [spec](https://amzn.github.io/ion-schema/docs/spec.html#core-types):
> If not specified, the default type is any.

This causes unexpected behavior in some situations. (See [`implicit_type_any_confusing_examples.isl`](https://gist.github.com/popematt/f98bff058327729f9deeecd946483a3d#file-implicit_type_any_confusing_examples-isl) for some examples.) In most cases, the default `type:any` is redundant because many of the constraints will already reject null values.

| Constraint | Allows null values? |
| ---- | ---- |
| `all_of` | Yes, if constraint arg is a list of nullable types
| `annotations` | Yes, always
| `any_of` | Yes, if constraint arg list has a nullable type
| `byte_length` | No
| `codepoint_length` | No
| `container_length` | No
| `contains` | No
| `element` | No
| `fields` | No
| `not` | Yes, unless constraint arg is a nullable type
| `one_of` | Yes, if constraint arg list has a nullable type
| `ordered_elements` | No
| `precision` | No
| `regex` | No
| `scale` | No
| `timestamp_offset` | No
| `timestamp_precision` | No
| `type` |  Yes, if constraint arg is a nullable type
| `valid_values` | Yes, if constraint arg includes a null value

There are some examples where the implicit `type:any` might be useful, but these edge cases are better off being dealt with by requiring a type definition to have at least one constraint. For example, the types of the fields (in the example below) might be a little confusing, but if type definitions were required to have at least one constraint, then one would simply add `type: any` or `type: $any` after the `occurs`.
```
type::{
  type: struct
  fields: {
    foo: { occurs: optional }
    bar: { occurs: required }
  }
}
```

**Proposed Solution**

Change the spec, starting in the next version of Ion Schema, so that there is no "default" `type` for type definitions. Type definitions will not be required to have a `type` constraint, but will be required to have at least one constraint.

**Comment by popematt** [comment]
An alternative solution that was brought up by @zslayton would be to require every type definition to have a `type:`. This also avoids the problem of unexpected/surprising behavior. However, we would need to consider if it makes certain idioms too verbose. For example:

```
// Any struct that is not annotated with Foo or Bar
// `type:` is always required
type::{
  name: not_foo_or_bar,
  type: struct,
  not: {
    type: struct
    any_of:[
      {
        type: struct,
        annotation: required::[foo]
      },
      {
        type: struct,
        annotation: required::[bar]
      } 
    ]
  }
}
```
Compare to:
```
// Any struct that is not annotated with Foo or Bar
// `type:` is not always required
type::{
  name: not_foo_or_bar,
  type: struct,
  not: {
    any_of:[
      { annotation: required::[foo] },
      { annotation: required::[bar] } 
    ]
  }
}

```

**Comment by popematt** [comment]
It was suggested offline by @almann that the default `type:` should be changed to `$any` (the top type). This is essentially identical to the original proposal in function, but it makes a lot more sense when explaining to someone who is learning Ion Schema.

**Comment by popematt** [comment]
Resolved by #76 in Ion Schema 2.0
